### ISSUE TITLE
Documentation/Recipe Request: Using redux-saga with TypeScript

### ISSUE THREAD

**Issue by huntwj** [redux-saga/redux-saga#1286]
I see that there are type definitions, and some people have gotten them to work. As someone that's new to TypeScript, however, converting my redux-saga to TypeScript has gone less than smoothly.

It would be nice if we could add a page of documentation (preferably written by someone that has had success in doing this...) outlining the process of converting JS sagas to TS and/or writing TS sagas from scratch.

Just in case my issue turns out to be easily solvable by the experts, I'll include it here:

Specifically, I am attempting to use `takeLatest(...)` and call my API request sagas, but I am getting type errors on those that take parameters.

My TypeKeys:
```
export enum TypeKeys {
  FETCH_REQUEST = 'Fetch Request',
  .
  .
  .
}
```

In my watchers generator:
```
    yield takeLatest(TypeKeys.FETCH_REQUEST, fetchRequestSaga);
```

and in my sagas:

```
export function *fetchRequestSaga(getState?: () => any): SagaIterator {
  .
  .
  .
}
```

Then the error is:
```
TS2345: Argument of type 'TypeKeys.FETCH_REQUEST' is not assignable to parameter of type 'Channel<(() => any) | undefined>'.
```

I used to be able to call takeLatest passing in a string as the action type, but things seem to change now that I'm dealing with a TS (string-based) enum.

**Comment by Andarist** [comment]
@aikoven do you know any good resources about the topic?

**Comment by aikoven** [comment]
@huntwj The error you see is a result of TypeScript's way of resolving overloaded signatures. `takeLatest` has two kinds of signatures: one for taking actions and another for taking from channels. For some reason TS failed to match your arguments with the first kind. Then it tried to match with the second kind, and of course it failed too, and returned an error for the last signature it tried.

Unfortunately, I don't know of a way to see the correct error here, except for commenting out all the signatures of `takeLatest` with channels in `node_modules/redux-saga/effects.d.ts`.

You may also try the new typings for the 1.0 (https://github.com/redux-saga/redux-saga/pull/1255), where these two groups of signatures are swapped, so that users are more likely to get the correct error, since `takeLatest` is more often used with actions than channels.

**Comment by erik-sn** [comment]
@huntwj did you ever find a resolution for this?

**Comment by huntwj** [comment]
I seem to have "moved past" the problems, though I cannot remember exactly what the solution was. I'm guessing it was related to using the 1.0 typings, but I can't remember off hand and unfortunately don't have time to do the necessary testing right now.

What I still think could be useful, and is something I may attempt to put together in the near future, would be a page of documentation providing a quick "getting started" guide for using redux saga with TypeScript.

**Comment by AleksandrChernyavenko** [comment]
I had error `Argument of type 'ChartActionTypes.FETCH_CHART' is not assignable to parameter of type 'Channel<{ payload: any; }>'.`

So problem was in `fetchChart(action: type)` - it has wrong type. I added type for action, and this fixed error 

```typescript
export function* fetchChart(action: FetchChartAction) { // <-- here added type FetchChartAction
  const {error, data} = yield call(fetchChartData, action)
  if (!error) {
    yield put(fetchChartSuccess(data))
  } else {
    yield put(fetchChartFailure(error))
  }
}

export function* watchFetchChart() {
  yield takeLatest(ChartActionTypes.FETCH_CHART, fetchChart) // <-- here pass `action` argument
}

```

**Comment by agzamovr** [comment]
If above solution didn't work for you (as it didn't for me) try this one https://github.com/redux-saga/redux-saga/issues/1188#issuecomment-425717322

**Comment by mynameistechno** [comment]
Are there any good examples/codebases that leverage redux-saga typescript types?

**Comment by gilbsgilbs** [comment]
I've been using Redux-Saga and TypeScript at work in a medium-sized front-end codebase for a while. Overall, it works fine and I want to share the main issues, pitfalls and solutions I used. If you find it interesting, it could be a starting point for writing a recipe.

## Setup

I defined a custom action interface that extends Redux's `Action` and that is inherited by all the actions I declare in my application. I think it is a best practice anyway. It somehow helps TypeScript at inferring types and avoids some explicit type casts:

```typescript
import { Action } from 'redux';

interface AppAction extends Action { 
    payload?: any;
}

interface LogAction extends AppAction {
  type: 'LOG_ACTION';
  payload: string;
}
```

Since TypeScript can guarantee there's no other `AppAction` having `LOG_ACTION` as type, you can then write:

```typescript
yield RSEffects.takeEvery(
  (action: CustomAction) => (
    action.type === 'LOG_ACTION'
    && action.payload === 'hello'
  ),
  someWorker,
);
```

instead of:

```typescript
yield RSEffects.takeEvery(
  (action: Action) => (
    action.type === 'LOG_ACTION'
    && (action as LogAction).payload === 'hello'
  ),
  someWorker,
);
```

The store can be instantiated normally :

```typescript
const sagaMiddleware = createSagaMiddleware();
const store = Redux.createStore(
    reducer,
    Redux.applyMiddleware(sagaMiddleware),
);
sagaMiddleware.run(rootSaga);
```

## Sagas

The simplest way to create a saga is to create a generator having `SagaIterator` as return type:

```typescript
import { SagaIterator } from '@redux-saga/core';
import { delay } from '@redux-saga/core/effects';

function* saga(): SagaIterator {
  yield delay(1000);
  console.log('ok');
}
```

Effects can be used as expected:

```typescript
function* saga(): SagaIterator {
    const chan = ReduxSaga.channel();  // <== Get Channel type inferred
    chan.put('hello');
    const res: string = yield RSEffects.take(chan);
    console.log(res);

    yield RSEffects.takeEvery(
        'LOG_ACTION',
        function* (action: LogAction): SagaIterator {
            console.log(action);
        },
    );
}
```

The `Saga` type can be used to reference sagas (sorry, I don't have a better example that comes into my mind yet, but I had to use the `Saga` type from time to time):

```typescript
import { Saga, SagaIterator } from '@redux-saga/core';
import { call, delay } from '@redux-saga/core/effects';

function* saga(subSaga: Saga): SagaIterator {
  yield delay(1000);
  yield call(subSaga);
}
```

## Return types

This is **THE** major painpoint when working with Redux-Saga and TypeScript: [the lack of support for generators return types](https://github.com/Microsoft/TypeScript/issues/2983). Hopefully, [TS team seem to have some plan on improving this for 3.5](https://github.com/Microsoft/TypeScript/issues/30555), but it's still unsure when it'll land and we'll have to update Redux-Saga types anyways. The only workaround is to type effects return types explicitly:

```typescript
function* saga(subSaga: Saga): SagaIterator {
  const response: Response = yield RSEffects.call(fetch, 'https://httpstat.us/200');  // <== Typed explicitely, otherwise we have "any"
  const txt: string = yield RSEffects.call([response, 'text']); // <== Same
  console.log(txt);
}
```

## Generic types and Sagas

This is the less straightforward part I had to deal with. TypeScript seems pretty bad at inferring generic types on generator functions. It may work in simple cases, but very often, TypeScript will report some typing errors when using the `call` effect on a Saga with generic types, or even worse, report no error at all even though typing is obviously incorrect. I still have not sorted out in which extent this is a TypeScript bug and/or limitation and/or Redux-Saga having improper typings. Still it bit me a few times.

Example:

```typescript
function* genericSaga<
  T extends (str: string) => number
>(callback: T): SagaIterator {
    yield RSEffects.delay(1000);
    console.log(callback('hello'));
}

function* saga: SagaIterator {
    yield RSEffects.call(genericSaga, 42); // WTF?
}
```

This is an example where TypeScript reports no error at all. Conversely, I had more complex cases I cannot really share where TypeScript would report an error although I would have expected it to have inferred a type. Luckily, it's not a very common use case. It's mainly useful for low-level primitives (e.g.: task queue that takes a worker as input parameter).

When working on this problem, I came across a workaround that I found elegant and I think is worth sharing. The solution was to refactor saga generators into plain effect functions. I found this pattern to be very much pleasant and convenient to use since it hides the `call` primitive to the calling saga. If there's no gotchas I missed, I think this pattern could be a recipe for untyped Redux-Saga and could be used consistently when declaring a saga that may be called by other sagas.

```typescript
function genericEffect<
  T extends (str: string) => number
>(callback: T): CallEffect {
    return call(function*(): SagaIterator {
        yield RSEffects.delay(1000);
        console.log(callback('hello'));
    });
}

function* saga: SagaIterator {
    yield genericEffect(42); // Error: 22 not assignable to Fn. This is expected.
    yield genericEffect((str) => str.length); // OK
}
```
---
Hope this helps. What do you think? Did I miss something important? Can it serve as a base for writing a recipe? Did I make any nasty mistake? If you have any question, do no hesitate.

**Comment by mynameistechno** [comment]
This is super helpful, thanks!

**Comment by nithincvpoyyil** [comment]
Hi All,

 I`m getting an error for following code, 

just followed [examples](https://redux-saga.js.org/docs/advanced/Testing.html).

```
const stack =[];
const generator = await runSaga({
      dispatch: (actionDispatched:any)=>{stack.push(actionDispatched)},
      getState: ()=>({data:[]})},
    }, fetchDataSaga).toPromise();
```

>  'await' expression is only allowed within an async function.ts(1308)'

 
Any suggestions ? thanks in advance.!!

**Comment by gilbsgilbs** [comment]
@nithincvpoyyil this is not a typing / typescript issue. As indicated, you need to find a way to use `await` keyword from an async function. If using jest, you can just make your test function async like this : https://jestjs.io/docs/en/asynchronous.html#async-await .

**Comment by nithincvpoyyil** [comment]
@gilbsgilbs : Thanks for the response. Yes, that was the issue. jest test function should be **async**, I will add it as a comment above. feeling so stupid :disappointed:

**Comment by jtomaszewski** [comment]
Seems like strong typing of generators is finally gonna come in TS 3.6? https://github.com/microsoft/TypeScript/issues/31639 Which means, we will be able to have well typed sagas using generic types? Are there some plans/work towards supporting it?

**Comment by Andarist** [comment]
Yes, when TS@3.6 gets released we are definitely going to work on improved typings.

**Comment by gilbsgilbs** [comment]
Not willing to spoil the party or anything, but https://github.com/microsoft/TypeScript/issues/31639 isn't actually sufficient to strongly type sagas and other `co`-style libraries as we would like. While it is indeed a slight improvement since we're now able to typecheck sagas return value:

```typescript
function* mySaga(): SagaIterator<number> {
  yield call(someEffect);
  return "not a number";  // This will fail
}
```

and even the `yield`/next assignment to some extent:

```typescript
function* mySaga(): SagaIterator<void, number> {
  const val = yield call(someEffect);  // val will have "number" type
}
```

but it remains too dumb to properly infer types in such cases:

```typescript
function* mySaga(): SagaIterator {
  return "not a number";  // This generator should have `string` ReturnType properly infered
}

function* anotherSaga(): SagaIterator {
  const val = yield call(mySaga);  // val will still have `any` type :(
}
```

And as far as I understand, there's nothing we can do about it at the moment. I'm looking for the actual issue to follow on TypeScript, but I'm unsure.

This isn't really related to this issue though. Maybe we could keep track of this in another issue @Andarist?

See https://github.com/microsoft/TypeScript/pull/30790#issuecomment-480467971 and https://github.com/Microsoft/TypeScript/issues/2983#issuecomment-230414026 for more information.

**Comment by gilbsgilbs** [comment]
I think one approach to improve the situation a bit would be to write [typescript-eslint](https://github.com/typescript-eslint/typescript-eslint) rules crafted for Saga that would for example ensure that:

- Every yield assignment is explicitly typed
- If the yielded effect is a CallEffect, ensure that the yield assignment type matches the return type of the CallEffect
- If the yield effect is a ForkEffect, ensure that the yield assignment type is a Task with a proper generic  type
- etc…

And generator return types could probably help to implement this more easily:

```typescript
function* childSaga(): SagaIterator<number> {
  yield delay(100);
  return 12;
}

function childCallEffect(): CallEffect<number> {
  return call(childSaga);
}

// until here, TS 3.6 should be enough to strongly type everything.
// starting from here, the linting rule would come into place

function* parentSaga(): SagaIterator<void> {
  // typeof call(childSaga) is CallEffect<number> => ensure we assign number here.
  const childSagaResult: number = yield call(childSaga);

  // And similarly
  const childCallEffectResult: number = yield childCallEffect();

  // typeof call(fetch<any>(…)) is CallEffect<Response<any>> => ensure we assign a Response<any> here.
  const reqResult: Response<any> = yield call(fetch<any>('http://example.com'));
}
```

I'm very unsure how realistic this idea is in term of complexity or even feasibility. Yet, some typescript-eslint rules are smart enough [to leverage type information](https://github.com/typescript-eslint/typescript-eslint/tree/c69f4b7132937f112d497213abb0b3fb69d85ae9#can-we-write-rules-which-leverage-type-information). Therefore, I believe it's feasible at very least, maybe just not worth the complexity.

The thing is I'm quite pessimistic on the ability of the TS team to come up with a general solution to this issue anytime soon. And I don't blame them, the general case seems very very hard to tackle after thinking of it a bit. Still, it may be more worth than we think to provide a very imperfect solution.

What do you think?

**Comment by Andarist** [comment]
If it's possible then it would be great if somebody could work on it - I'm afraid the effort would be community-driven though, as personally I don't have time to explore this right now.

**Comment by gilbsgilbs** [comment]
I made some experiments tonight and it appears to be feasible. I updated redux-saga types to allow specifying a return type to saga generators (using TS 3.6 generator return types) and I am able to get this return type within an ESLint rule :tada: . I may open a PR sometime for Sagas return types and if it makes it, I may release an ESLint plugin to enforce strong types on effects. We could then write a good Typescript recipe :) .

**Comment by Andarist** [comment]
Great, looking forward to it.

**Comment by dwilt** [comment]
@gilbsgilbs any progress on this?

**Comment by gilbsgilbs** [comment]
@dwilt progress on the ESLint rule? I don't think I'll spend any time on it because somebody found a [simpler solution](https://github.com/redux-saga/redux-saga/issues/884#issuecomment-513200739) that would cope well with type inference (at the price of slight changes in the way sagas are invoked). If our community don't fall into pointless quarrels waged by ego, I think it might be the definitive solution to strongly typed sagas.

About the improved typings for redux-saga with TS3.6 (which I think is a prerequisite for all the solutions we have), I made a WIP PR (#1892) however, [as I mentioned](https://github.com/redux-saga/redux-saga/pull/1892#issuecomment-515794720), I'm stuck and I need to figure out why `typesVersions` doesn't work as I would expect. Not sure when I will find enough patience and time to dig into this, but probably not soon since I'm very prone to rage quitting on packaging-related topics :sweat: . If you're willing to help making this happen sooner rather than later, I think this issue is the very next step :wink: .

**Comment by neurosnap** [comment]
I’d like to overhaul the docs to emphasize Typescript in the near future.  Currently we are blocked by TS itself providing better support for yield overrides.  You can track that progress here: https://github.com/microsoft/TypeScript/issues/43632

Going to close this issue for now.
